R/conformal_infer.R

```diff
@@ -0,0 +1,455 @@
+
+  pkg_inst <- lapply(ctrl$required_pkgs, rlang::is_installed)
+  pred_val <- new_data$.pred
+  bound <- new_data$.bound
+  new_data <- dplyr::select(new_data, -.pred, -.bound)
+  new_data[[y_name]] <- NA_real_
+  trial_data <- dplyr::bind_rows(train_data, new_data)
+
+  trial_vals <- seq(pred_val - bound, pred_val + bound, length.out = ctrl$trial_points)
+
+  res <-
+    purrr::map_dfr(
+      trial_vals,
+      ~ trial_fit(.x,
+                  trial_data = trial_data,
+                  wflow = model,
+                  level = level)
+    )
+
+  compute_bound(res, pred_val)
+}
+
+compute_bound <- function(x, predicted) {
+  x <- x[stats::complete.cases(x),]
+  if (all(x$difference < 0)) {
+    rlang::warn("Could not determine bounds.")
+    res <-
+      dplyr::tibble(.pred_lower = NA_real_,
+                    .pred = predicted,
+                    .pred_upper = NA_real_)
+    return(res)
+  }
+
+  upper <- x[x$trial >= predicted & x$difference >= 0,]
+  if (nrow(upper) > 0) {
+    upper <- min(upper$trial)
+  } else {
+    upper <- NA_real_
+  }
+  lower <- x[x$trial <= predicted & x$difference >= 0,]
+  if (nrow(lower) > 0) {
+    lower <- max(lower$trial)
+  } else {
+    lower <- NA_real_
+  }
+  dplyr::tibble(.pred_lower = lower, .pred_upper = upper)
+}
+
+# ------------------------------------------------------------------------------
+
+optimize_all <- function(new_data, model, train_data, level, ctrl) {
+  # purrr::map_dfr(
+  furrr::future_map_dfr(
+    new_data,
+    optimize_one,
+    model = model,
+    train_data = train_data,
+    level = level,
+    ctrl,
+    .progress = ctrl$progress,
+    .options = furrr::furrr_options(seed = ctrl$seed, stdout = FALSE)
+  )
+}
+
+get_diff <- function(val, ...) {
+  res <- trial_fit(val, ...)
+  res$difference
+}
+
+optimize_one <- function(new_data, model, train_data, level, ctrl) {
+  y_name <- get_outcome_name(model)
+
+  pkg_inst <- lapply(ctrl$required_pkgs, rlang::is_installed)
+  pred_val <- new_data$.pred
+  bound <- new_data$.bound
+  new_data <- dplyr::select(new_data, -.pred, -.bound)
+  new_data[[y_name]] <- NA_real_
+  trial_data <- dplyr::bind_rows(train_data, new_data)
+
+  upper <-
+    try(
+      stats::uniroot(get_diff, c(pred_val, pred_val + bound),
+                     maxiter = ctrl$max_iter, tol = ctrl$tolerance,
+                     extendInt = "upX",
+                     trial_data, model, level),
+      silent = TRUE)
+
+  lower <-
+    try(
+      stats::uniroot(get_diff, c(pred_val - bound, pred_val),
+                     maxiter = ctrl$max_iter, tol = ctrl$tolerance,
+                     extendInt = "downX",
+                     trial_data, model, level),
+      silent = TRUE)
+
+  dplyr::tibble(
+    .pred_lower = get_root(lower, ctrl),
+    .pred_upper = get_root(upper, ctrl)
+  )
+}
+
+get_root <- function(x, ctrl) {
+  if (inherits(x, "try-error")) {
+    rlang::warn("Could not finish the search process due to errors.")
```

Comment:
This is another place where I'd wonder if it's helpful for us to pass along the condition message.